README
ActsAsGold
==========

Allows a model to have copper, silver and gold, based on a single integer column on the model itself. 

For example, this Character model has the following migration:

add_column :money, :integer, :limit => 20 


Example
=======

Example goes here.


Copyright (c) 2008 [name of plugin creator], released under the MIT license


Changelog
2008-08-14
 * [1.0.5] ariejan - Fixed bug with gem not monkey patching ActiveRecord::Base correctly. 
 * [1.0.4] ariejan - Bump to 1.0.4 to ping github

2008-08-13
 * [1.0.3] ariejan - Bump to 1.0.3 to ping github
 * [1.0.2] ariejan - Added tests and fixed a bug when using a custom column to store the money value.
 * [1.0.1] ariejan - Minor changes to the gemspec.
 * [1.0.0] ariejan - Initial release of the acts_as_gold plugin and gem..


test/acts_as_gold_test.rb
require 'test/unit'

require 'rubygems'
gem 'activerecord', '>= 1.15.4.7794'
require 'active_record'

require "#{File.dirname(__FILE__)}/../init"

ActiveRecord::Base.establish_connection(:adapter => "sqlite3", :dbfile => ":memory:")

def setup_db
  ActiveRecord::Schema.define(:version => 1) do
    create_table :players do |t|
      t.column :money, :integer
    end
    
    create_table :another_players do |t|
      t.column :pennies, :integer
    end
  end
end

def teardown_db
  ActiveRecord::Base.connection.tables.each do |table|
    ActiveRecord::Base.connection.drop_table(table)
  end
end

class Player < ActiveRecord::Base
  acts_as_gold # defaults to :column => :money
end

class AnotherPlayer < ActiveRecord::Base
  acts_as_gold :column => :pennies
end

class ActsAsGoldTest < Test::Unit::TestCase
  
  def setup
    # Creates a Player iwth 250G, 44S and 55C
    setup_db
    Player.create(:money => 2504455)
    @player = Player.find(:first)
  end
  
  def teardown
    teardown_db
  end
  
  def test_alternate_column_name
    AnotherPlayer.create(:pennies => 1005095)
    @alt_player = AnotherPlayer.find(:first)
    
    assert_equal 1005095, @alt_player.pennies
    assert_equal 100, @alt_player.gold
    assert_equal 50, @alt_player.silver
    assert_equal 95, @alt_player.copper
  end
  
  def test_money_retrieval
    assert_equal 2504455, @player.money
    
    assert_equal 250, @player.gold
    assert_equal 44, @player.silver
    assert_equal 55, @player.copper
  end
  
  def test_numeric_extensions
    assert_equal 2500000, 250.gold
    assert_equal 2500, 25.silver
    assert_equal 25, 25.copper
  end
  
  def test_money_earning
    assert_equal 2504455, @player.money
    assert @player.earn(3.gold + 22.silver + 33.copper)
    assert_equal 2536688, @player.money
  end
  
  def test_money_spending
    assert_equal 2504455, @player.money
    assert @player.spend(3.gold + 22.silver + 33.copper)
    assert_equal 2472222, @player.money
  end
  
  def test_too_much_money_spending
    assert_equal 2504455, @player.money
    assert_raise(ActsAsGold::NotEnoughMoneyError) { @player.spend(300.gold) }
    assert_equal 2504455, @player.money
  end
end


lib/acts_as_gold.rb
module ActsAsGold
  # Thrown when trying to spend more than you have.
  class NotEnoughMoneyError < StandardError
  end
  
  def self.included(base)
    base.extend(ClassMethods)
  end

  # This +acts_as+ extension provides the capabilities for splitting a given interger column into 
  # three separte coin values: Copper, Silver and Gold. 
  #
  # A simple example: a Charachter has money ( t.integer :money, :limit => 20)
  #
  #   class Character < ActiveRecord::Base
  #     acts_as_gold                      # Uses the money attributes
  #     acts_as_gold :column => :pennies  # Uses the pennies attributes
  #   end
  #
  #   character.money   = 57503
  #   character.gold    = 5
  #   character.silver  = 75
  #   character.copper  = 3
  #
  #   character.money  += 2.gold + 10.copper
  #   character.gold   = 7
  #   character.copper = 13
  #   character.money  = 77513
  #
  # Copper and Silver have a maximum of 99. E.g. 100 copper => 1 silver and 100 silver => 1 Gold. 
  # The maximum amount of money depends on the integer type used in the database:
  #
  # signed int(11)             2,147,483,647   =>  214,748 Gold, 36 Silver, 47 Copper
  # signed int(20) 9,223,372,036,854,775,807   =>  922,337,203,685,477 Gold, ++8 Silver, 07 Copper
  #
  module ClassMethods
    # Configuration options are:
    #
    # * +column+ - specifies the column name to use for keeping the money integer (default: +money+)
    def acts_as_gold(options = {})
      configuration = { :column => "money" }
      configuration.update(options) if options.is_a?(Hash)

      class_eval <<-EOV
        include ActsAsGold::InstanceMethods
        
        def money_column
          '#{configuration[:column]}'
        end

      EOV
    end
  end
  
  # Allow Fixnum and Bignum to easily convert to Gold, Silver or Copper. 
  # This allows for things like:
  #
  # character.money = 2.gold + 45.silver + 50.copper
  module IntegerExtensions
    # 1.gold => 10000
    def gold
      self * 10000
    end
    
    # 1.silver => 100
    def silver
      self * 100
    end
    
    # Dummy, 1.copper => !
    def copper
      self
    end
  end

  # All the methods available to records that have the acts_as_gold method enabled.
  module InstanceMethods
    # Earn money. 
    #
    # Either enter a total money value, or sum it up with gold and def silver. To earn
    # 1 Gold, 95 silver and 0 copper you can do the following:
    #
    # character.earn(19500)
    # character.earn(1.gold + 95.silver)
    #
    # This return true if the amount was added successfully
    def earn(amount)
      update_attribute(:money, money + amount)
    end
    
    # Spend money
    #
    # You can specify money the same way as with earning money.
    #
    # We return true if the money was spend successfully.
    #
    # This will raise a 'ActiveRecord::Acts::Gold::NotEnoughMoneyError' when there's not enough money
    def spend(amount)
      if money >= amount
        update_attribute(:money, money - amount)
      else
        raise ActsAsGold::NotEnoughMoneyError
      end
    end
    
    # Return the amount of Gold
    def gold
      split_copper.first.divmod(100).first
    end

    def silver
      split_copper.first.divmod(100).last
    end
    
    def copper
      split_copper.last
    end

    private
    
    def split_copper
      self.send(money_column).divmod(100)
    end
  end 
end

class Fixnum
  include ActsAsGold::IntegerExtensions
end

class Bignum
  include ActsAsGold::IntegerExtensions
end

ActiveRecord::Base.class_eval { include ActsAsGold }

lib/active_record/acts/gold.rb
module ActiveRecord  
  module Acts #:nodoc:
    module Gold #:nodoc:
      def self.included(base)
        base.extend(ClassMethods)
      end

      # This +acts_as+ extension provides the capabilities for splitting a given interger column into 
      # three separte coin values: Copper, Silver and Gold. 
      #
      # A simple example: a Charachter has money ( t.integer :money, :limit => 20)
      #
      #   class Character < ActiveRecord::Base
      #     acts_as_gold :money
      #   end
      #
      #   character.money   = 57503
      #   character.gold    = 5
      #   character.silver  = 75
      #   character.copper  = 3
      #
      #   character.money  += 2.gold + 10.copper
      #   character.gold   = 7
      #   character.copper = 13
      #   character.money  = 77513
      #
      # Copper and Silver have a maximum of 99. E.g. 100 copper => 1 silver and 100 silver => 1 Gold. 
      # The maximum amount of money depends on the integer type used in the database:
      #
      # signed int(11)             2,147,483,647   =>  214,748 Gold, 36 Silver, 47 Copper
      # signed int(20) 9,223,372,036,854,775,807   =>  922,337,203,685,477 Gold, 58 Silver, 07 Copper
      #
      module ClassMethods
        # Configuration options are:
        #
        # * +column+ - specifies the column name to use for keeping the money integer (default: +money+)
        def acts_as_gold(options = {})
          configuration = { :column => "money" }
          configuration.update(options) if options.is_a?(Hash)

          class_eval <<-EOV
            include ActiveRecord::Acts::Gold::InstanceMethods
            
            def money_column
              '#{configuration[:column]}'
            end

          EOV
        end
      end
      
      # Allow Fixnum and Bignum to easily convert to Gold, Silver or Copper. 
      # This allows for things like:
      #
      # character.money = 2.gold + 45.silver + 50.copper
      module IntegerExtensions
        # 1.gold => 10000
        def gold
          self * 10000
        end
        
        # 1.silver => 100
        def silver
          self * 100
        end
        
        # Dummy, 1.copper => !
        def copper
          self
        end
      end

      # All the methods available to records that have the acts_as_gold method enabled.
      module InstanceMethods
        # Earn money. 
        #
        # Either enter a total money value, or sum it up with gold and def silver. To earn
        # 1 Gold, 95 silver and 0 copper you can do the following:
        #
        # character.earn(19500)
        # character.earn(1.gold + 95.silver)
        #
        # This return true if the amount was added successfully
        def earn(amount)
          update_attribute(:money, money + amount)
        end
        
        # Spend money
        #
        # You can specify money the same way as with earning money.
        #
        # We return true if the money was spend successfully.
        #
        # This will raise a 'ActiveRecord::Acts::Gold::NotEnoughMoneyError' when there's not enough money
        def spend(amount)
          if money >= amount
            update_attribute(:money, money - amount)
          else
            raise HotChocolate::NotEnoughMoneyError
          end
        end
        
        # Return the amount of Gold
        def gold
          split_copper.first.divmod(100).first
        end

        def silver
          split_copper.first.divmod(100).last
        end
        
        def copper
          split_copper.last
        end

        private
        
        def split_copper
          self.send(money_column).divmod(100)
        end
      end 
    end
  end
end

class Fixnum
  include ActiveRecord::Acts::Gold::IntegerExtensions
end

class Bignum
  include ActiveRecord::Acts::Gold::IntegerExtensions
end